### PR TITLE
[DEV-2987] Implement Spark Thrift and Databricks FEAST offline store datasource

### PR DESCRIPTION
**Pull request by kchua78**
## Description

Implement Spark Thrift and Databricks FEAST offline store datasource for online serving support

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/feast/infra/offline_stores/spark_thrift.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+        Parameters
+        ----------
+        timeout: Optional[int]
+            Timeout in seconds
+
+        Returns
+        -------
+        pd.DataFrame
+        """
+        statements = self.query.split("---EOS---")
```
Comment (YS-L) [inline featurebyte/feast/infra/offline_stores/spark_thrift.py:277]: Is the query here generated by us? Does it contain the `---EOS---` string?

File: featurebyte/feast/infra/offline_stores/spark_thrift_source.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        """
+        spark_thrift_options_proto = DataSourceProto.CustomSourceOptions(
+            configuration=json.dumps(
+                {"catalog": self.catalog, "schema": self.schema, "table": self.table}
+            ).encode()
+        )
+        return spark_thrift_options_proto
+
+
+class SavedDatasetSparkThriftStorage(SavedDatasetStorage):
```
Comment (YS-L) [inline featurebyte/feast/infra/offline_stores/spark_thrift_source.py:202]: What is the purpose of this class? Do we need the same for databricks?

File: featurebyte/feast/infra/offline_stores/spark_thrift.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        return SparkSession(**self.dict())
+
+
+class SparkThriftOfflineStore(OfflineStore):
+    """
+    Offline store for Spark Thrift Server
+    """
+
+    @staticmethod
+    def pull_latest_from_table_or_query(
```
Comment (YS-L) [inline featurebyte/feast/infra/offline_stores/spark_thrift.py:105]: Curious what are the differences between the query generated by this vs the spark offline store? https://github.com/feast-dev/feast/blob/9a3590ea771ca3c3224f5e1a833453144e54284e/sdk/python/feast/infra/offline_stores/contrib/spark_offline_store/spark.py#L59-L62

Comment (YS-L) [reply featurebyte/feast/infra/offline_stores/spark_thrift.py:105]: Oh it's the dependency on pyspark. The query probably is quite similar.

File: featurebyte/feast/infra/offline_stores/spark_thrift.py (hunk 1 above)
Comment (kchua78) [reply featurebyte/feast/infra/offline_stores/spark_thrift.py:277]: Oh indeed. I forgot to change this part - we won't generate multi-line statements

File: featurebyte/feast/infra/offline_stores/spark_thrift.py (hunk 3 above)
Comment (kchua78) [reply featurebyte/feast/infra/offline_stores/spark_thrift.py:105]: Yes the query is derived from there but implemented using sqlglot for better robustness and to ensure proper quoting

File: featurebyte/feast/infra/offline_stores/spark_thrift_source.py (hunk 2 above)
Comment (kchua78) [reply featurebyte/feast/infra/offline_stores/spark_thrift_source.py:202]: It does not seem to be used in our serving implementation - will remove.

We are using the same data source class (SparkThriftSource) for spark, databricks and databricks unity, the difference is only in the offline store configs where each offline store type uses a different implementation for establishing a thrift session

File: featurebyte/feast/infra/offline_stores/spark_thrift_source.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        spark_thrift_options_proto = DataSourceProto.CustomSourceOptions(
+            configuration=json.dumps(
+                {"catalog": self.catalog, "schema": self.schema, "table": self.table}
+            ).encode()
+        )
+        return spark_thrift_options_proto
+
+
+#
+# class SavedDatasetSparkThriftStorage(SavedDatasetStorage):
```
Comment (YS-L) [inline featurebyte/feast/infra/offline_stores/spark_thrift_source.py:201]: Should we just remove this if not needed?

**Review by YS-L — APPROVED** [review]

**Pull request merged**